orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -194,43 +209,133 @@ Status FromTensorProtos(
 
   constexpr size_t m_bytes = 1024 * 1024;
   size_t fbs_buffer_size = 0U;
+  size_t fbs_potential_external_buffer_size = 0;
   for (const auto& tensor_proto : trainable_tensor_protos) {
-    fbs_buffer_size += tensor_proto.ByteSizeLong();
+    auto bytes = tensor_proto.ByteSizeLong();
+    fbs_buffer_size += bytes;
+
+    if (tensor_proto.data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
+      std::unique_ptr<ExternalDataInfo> external_data_info;
+      ORT_RETURN_IF_ERROR(ExternalDataInfo::Create(tensor_proto.external_data(), external_data_info));
+      fbs_buffer_size += external_data_info->GetLength();
+    }
+
+    if (bytes > onnxruntime::fbs::utils::kMinimumSizeForExternalData) {  // assuming strings aren't trainable
+      fbs_potential_external_buffer_size += bytes;
+    }
   }
+
   for (const auto& tensor_proto : non_trainable_tensor_protos) {
-    fbs_buffer_size += tensor_proto.ByteSizeLong();
+    auto bytes = tensor_proto.ByteSizeLong();
+    fbs_buffer_size += bytes;
+
+    if (tensor_proto.data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
+      ORT_RETURN_IF_NOT(tensor_proto.external_data()[2].key() == "length",
+                        "Invalid external data for ", tensor_proto.name());
+      // external_data field is a dictionary of strings
+      // length is stored as a string
+      fbs_buffer_size += std::stoull(tensor_proto.external_data()[2].value());
+    }
+
+    if (bytes > onnxruntime::fbs::utils::kMinimumSizeForExternalData &&
+        tensor_proto.data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING) {
+      fbs_potential_external_buffer_size += bytes;
+    }
   }
 
   // Align buffer size to 1MB.
+  // TODO: This should probably also have an allowance for the other parts of the tensor if we're trying to avoid
+  // all reallocs during serialization. e.g. name, shape, data type, offset if external data is used.
   fbs_buffer_size = std::max(fbs_buffer_size, m_bytes);
   fbs_buffer_size = ((fbs_buffer_size + m_bytes - 1) / m_bytes) * m_bytes;
+
+  // 2GB is the real limit, but fbs_buffer_size doesn't include space for initializer names and shapes.
+  // for now, arbitrarily use ~1.8GB as a conservative 'safe' value with a 10% buffer.
+  const bool use_external_data = fbs_buffer_size >= external_data_threshold * m_bytes;
+
+  fbs::utils::ExternalDataWriter external_data_writer = nullptr;
+  std::optional<std::ofstream> external_data_stream;
+
+  if (use_external_data) {
+    // 2GB. Have to use external file for data.
```

Comment:
Now that the threshold is passed in these comments aren't relevant here. Could maybe move the into on using 1.8GB vs 2GB to the comments in the header for SaveCheckpoint where the default is set.